Decompiles Remaining `rwrp` Function & Data

hohle opened the pull request:
Adds an a PSX implementation of `EntityRWarpRoom`. This is similar to `EntityWarpRoom`, but needs to raise and lower the player stands on, and obviously has to warp to other inverted rooms.

This also pulls in the remaining data and static storage, completing the `rwrp` mapping.

hohle commented on src/st/rwrp/warp.c:
```diff
@@ -1,13 +1,351 @@
 #include "rwrp.h"
+#include "common.h"
+
+#define SE_WARP_ENTER 0x636
```
Reused an existing define in the `rwrp` header.

hohle commented on src/st/rwrp/warp.c:
```diff
@@ -1,13 +1,351 @@
+        // Wait for player to press the UP button
+        if (collision & 0x4 && (g_pads[0].pressed & PAD_UP) &&
+            UNDEFINED(GetDistanceToPlayerX)() < 8 &&
+            !(g_Player.unk0C & 0xC5CF3EF7)) {
```
defined a mask.

hohle commented on src/st/rwrp/warp.c:
```diff
@@ -1,13 +1,351 @@
+            D_80097488.y.i.hi += 1;
+
+            if (g_pads[0].pressed & PAD_UP &&
+                UNDEFINED(GetDistanceToPlayerX)() < 8 &&
```
I added the `UNDEFINED` macro because I've seen this pattern several times now. In the past I've written a specific cast or macro for the instance being undefined or cast the function inline, for example: https://github.com/Xeeynamo/sotn-decomp/blob/8cd4eb91bc50dbafda0618f45972fbe41eb489ba/src/st/st0/30030.c#L111

I'm happy to change the name, which may not reflect the intent.

Xeeynamo replied:
Having a simple function prototype will work just fine. I find the `UNDEFINED` confusing.

sozud commented:
Did you push? I'm not seeing any changes

hohle commented:
Just pushed.

Xeeynamo commented on src/st/rwrp/warp.c:
```diff
@@ -1,12 +1,366 @@
+//
+// Value: 0xC5CF3EF7
+//
+// TODO: this value is also used in src/st/wrp/warp.c
```
Instead of leaving here a TODO, do you want to copy & paste this in `warp.c`?

Xeeynamo reviewed (commented):
This is great!

hohle commented:
The nice thing about the wrapper macro is that it can be easily nullified for PC or other builds where it doesn't make sense.

Xeeynamo commented:
> The nice thing about the wrapper macro is that it can be easily nullified for PC or other builds where it doesn't make sense.

I do not understand the practical use of it?

hohle commented:
I've renamed the macro again 😜 and updated the comment.

Our compiler settings (wisely) disallow use of functions which are missing their declaration. This wasn't the case with GCC 2.6 and wasn't the default until the C99 standard. I'm trying to bridge the gap between loose declarations and stricter warnings and defaults with modern compilers.

There are several instances where it appears the original source was missing a function declaration, so the code was compiled using the [implicit K&R function signature](https://www.gnu.org/software/c-intro-and-ref/manual/html_node/Old_002dStyle-Function-Definitions.html) where the return value is assumed to be an `int` and all parameters are assumed to be `int` sized. To get an exact match on the binary, the function needs to be called with an implicit signature, but since that function is already declared with the correct signature, the generated binary would not match.

To get a binary match, the function needs to be cast to an `int(*)()`, which I've previously done by declaring a one-off macro of that specific function (in this case, it would be something like `#define fake_GetDistanceToPlayerX ((int(*)())GetDistanceToPlayerX)`. This should be modified to be conditional for the PC version to just call the original function since the signature change is not required.

Instead of doing all of that, wrapping the function name at the call sight with `IMPLICIT` will perform the cast on original architectures, but avoid it on PC.

I've update `src/st/st0/30030.c` as well to use the macro for illustration and can pull this out into a separate commit to decouple the change.

Xeeynamo commented:
> I've renamed the macro again 😜 and updated the comment.
> 
> Our compiler settings (wisely) disallow use of functions which are missing their declaration. This wasn't the case with GCC 2.6 and wasn't the default until the C99 standard. I'm trying to bridge the gap between loose declarations and stricter warnings and defaults with modern compilers.
> 
> There are several instances where it appears the original source was missing a function declaration, so the code was compiled using the [implicit K&R function signature](https://www.gnu.org/software/c-intro-and-ref/manual/html_node/Old_002dStyle-Function-Definitions.html) where the return value is assumed to be an `int` and all parameters are assumed to be `int` sized. To get an exact match on the binary, the function needs to be called with an implicit signature, but since that function is already declared with the correct signature, the generated binary would not match.
> 
> To get a binary match, the function needs to be cast to an `int(*)()`, which I've previously done by declaring a one-off macro of that specific function (in this case, it would be something like `#define fake_GetDistanceToPlayerX ((int(*)())GetDistanceToPlayerX)`. This should be modified to be conditional for the PC version to just call the original function since the signature change is not required.
> 
> Instead of doing all of that, wrapping the function name at the call sight with `IMPLICIT` will perform the cast on original architectures, but avoid it on PC.
> 
> I've update `src/st/st0/30030.c` as well to use the macro for illustration and can pull this out into a separate commit to decouple the change.

Yep I get why you created that macro. But I still think it is largely unnecessary. In case of `src/st/st0/30030.c` we just need to split the C files properly in the same way we're doing with WRP, NP3 and NZ0 to then allow GCC 2.6.x to use the implicit signature. I am not worried about a PC version as we're not there yet.

hohle commented:
`IMPLICIT` removed.
